homeassistant/helpers/icon.py

```diff
@@ -1,7 +1,165 @@
+from homeassistant.util.json import load_json_object
+
+from .translation import build_resources
+
+ICON_LOAD_LOCK = "icon_load_lock"
+ICON_CACHE = "icon_cache"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def _component_icons_path(component: str, integration: Integration) -> str | None:
+    """Return the icons json file location for a component.
+
+    Ex: components/hue/icons.json
+    If component is just a single file, will return None.
+    """
+    domain = component.rpartition(".")[-1]
+
+    # If it's a component that is just one file, we don't support icons
+    # Example custom_components/my_component.py
+    if integration.file_path.name != domain:
+        return None
+
+    return str(integration.file_path / "icons.json")
+
+
+def _load_icons_files(icons_files: dict[str, str]) -> dict[str, dict[str, Any]]:
+    """Load and parse icons.json files."""
+    return {
+        component: load_json_object(icons_file)
+        for component, icons_file in icons_files.items()
+    }
+
+
+async def _async_get_component_icons(
+    hass: HomeAssistant,
+    components: set[str],
+    integrations: dict[str, Integration],
+) -> dict[str, Any]:
+    """Load icons."""
+    icons: dict[str, Any] = {}
+
+    # Determine files to load
+    files_to_load = {}
+    for loaded in components:
+        domain = loaded.rpartition(".")[-1]
+        if (path := _component_icons_path(loaded, integrations[domain])) is None:
+            icons[loaded] = {}
+        else:
+            files_to_load[loaded] = path
+
+    # Load files
+    if files_to_load and (
+        load_icons_job := hass.async_add_executor_job(_load_icons_files, files_to_load)
+    ):
+        icons |= await load_icons_job
+
+    return icons
+
+
+class _IconsCache:
+    """Cache for icons."""
+
+    __slots__ = ("_hass", "_loaded", "_cache")
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the cache."""
+        self._hass = hass
+        self._loaded: set[str] = set()
+        self._cache: dict[str, dict[str, Any]] = {}
+
+    async def async_fetch(
+        self,
+        category: str,
+        components: set[str],
+    ) -> dict[str, dict[str, Any]]:
+        """Load resources into the cache."""
+        if components_to_load := components - self._loaded:
+            await self._async_load(components_to_load)
+
+        return {
+            component: result
+            for component in components
+            if (result := self._cache.get(category, {}).get(component))
+        }
+
+    async def _async_load(self, components: set[str]) -> None:
+        """Populate the cache for a given set of components."""
+        _LOGGER.debug(
+            "Cache miss for: %s",
+            ", ".join(components),
+        )
+
+        integrations: dict[str, Integration] = {}
+        domains = list({loaded.rpartition(".")[-1] for loaded in components})
+        ints_or_excs = await async_get_integrations(self._hass, domains)
```

Comment:
I think we should write out the words in the variable name. `ints` sounds like integers to me.